### ISSUE TITLE
Allow Advanced PathSpec optional behavior in ServletContextHandler addServlet and addFilter

### ISSUE THREAD

**Issue by joakime** [eclipse/jetty.project#5810]
**Jetty version**
Jetty 9.4.x

**Description**
Currently the `ServletHandler` uses the `PathMappings` feature of Jetty internals.

And it (correctly) only adds `ServletPathSpec` in response to the web descriptor or embedded usage
of ServletContextHandler.addServlet() or ServletContextHandler.addFilter() usage.

This is a request to add the ability to add servlets or filters based on one of the more advanced PathSpec features
Such as RegexPathSpec or UriTemplatePathSpec.

Ideally supporting :
* Set this from a ServletContextHandler or WebAppContext via initialized PathSpec object.
* Set from web descriptor (or annotations) via string with special keywords to break out of the Servlet behaviors.
  * Example: `@WebServlet("regex:/a/[^/]+/b/[^/]+")` or `<url-pattern>uri:/a/{foo}/b/{zed}</url-pattern>`

**Comment by gregw** [comment]
Two approaches: 
  1) allows arbitrary types of path spec to be passed in as Strings.  This would need less plumbing changes and would allow such patterns to come from more sources.  But we'd need a robust way of converting from such an arbitrary string to a specific pathSpec type.  Also would need optional validation so that we could enforce normal pathspecs only from webapps.
  2) add API that allows `PathSpec`s to be passed in as an instance rather than as a string.  Thus the caller will set the type.  Will need more changes to the plumbing and will only be available from the new API.

I think 2) is the best approach if it matches the requirements.

**Comment by janbartel** [comment]
Late to the party here, but I just want to note that I am **not** in favour of putting any extended path syntax into web.xml or annotations.

**Comment by joakime** [comment]
> Late to the party here, but I just want to note that I am **not** in favour of putting any extended path syntax into web.xml or annotations.

I agree with this.
IMO, Leaving this feature for only embedded-jetty users and `WEB-INF/jetty-web.xml` users are the only acceptable solutions.

**Comment by gregw** [comment]
So I have changed the PR so that rather than an extra non-standard `PathSpec` plus a String[] of stnadard path specs, all mappings are now stored as `PathSpec[]`, including constraints, filters and servlets.
The string based methods remain, but only work on `ServletPathSpec`.  Using the `PathSpec` API is the only way to add arbitrary path spec types.

I've added some testing, but more is probably needed.   Specially for the complex way security constraints are merged.
